radsat-sk/src/main.c

```diff
@@ -7,104 +7,376 @@
+/** Mission Init Task Priority. Does initializations that need to be ran post-scheduler; low priority task. */
+static const int missionInitTaskPriority = configMAX_PRIORITIES - 4;
+
+
+/***************************************************************************************************
+                                       PRIVATE FUNCTION STUBS
+***************************************************************************************************/
+
+static void initBoard(void);
+static int initDrivers(void);
+static int initTime(void);
+static int initSubsystems(void);
+
+static int initObcWatchdog(void);
+static int initMissionTasks(void);
+
+void MissionInitTask(void* parameters);
+
+
+/***************************************************************************************************
+                                                MAIN
+***************************************************************************************************/
+
+/**
+ * The main application entry point.
+ * @return Never returns.
+ */
+int main(void) {
+
+	// initialize internal OBC board settings
+	initBoard();
+
+	// initialize the Hardware Abstraction Library (HAL) drivers
+	initDrivers();
+
+	// initialize external components and the Satellite Subsystem Interface (SSI)
+	initSubsystems();
+
+	// initialize the internal OBC watchdog, and start a task that automatically pets it
+	initObcWatchdog();
+
 #ifdef TEST
-#include <RTestSuite.h>
-#endif /* TEST */
-
-#define ENABLE_MAIN_TRACES 1
-#if ENABLE_MAIN_TRACES
-	#define MAIN_TRACE_INFO			TRACE_INFO
-	#define MAIN_TRACE_DEBUG		TRACE_DEBUG
-	#define MAIN_TRACE_WARNING		TRACE_WARNING
-	#define MAIN_TRACE_ERROR		TRACE_ERROR
-	#define MAIN_TRACE_FATAL		TRACE_FATAL
-#else
-	#define MAIN_TRACE_INFO(...)	{ }
-	#define MAIN_TRACE_DEBUG(...)	{ }
-	#define MAIN_TRACE_WARNING(...)	{ }
-	#define MAIN_TRACE_ERROR		TRACE_ERROR
-	#define MAIN_TRACE_FATAL		TRACE_FATAL
-#endif
 
+	// TODO: run tests
 
-void taskMain(void* parameters)
-{
-	(void)parameters;
+#else	/* TEST */
 
-	WDT_startWatchdogKickTask(10 / portTICK_RATE_MS, FALSE);
+	// TODO: Antenna Diagnostic & Deployment (if necessary)
 
-	while(1) {
-		LED_wave(1);
-		LED_waveReverse(1);
-		LED_wave(1);
-		LED_waveReverse(1);
+	// TODO: Satellite Diagnostic Check (if applicable - may be done later instead)
 
-		vTaskDelay(500 / portTICK_RATE_MS);
-	}
+	// initialize the Mission Initialization Task
+	int error = xTaskCreate(MissionInitTask,
+							(const signed char*)"Mission Initialization Task",
+							DEFAULT_TASK_STACK_SIZE,
+							NULL,
+							missionInitTaskPriority,
+							&missionInitTaskHandle);
+
+	if (error != pdPASS)
+		debugPrint("main(): failed to create MissionInitTask.\n");
+
+#endif	/* TEST */
+
+	// start the FreeRTOS Scheduler - NEVER GETS PAST THIS LINE
+	vTaskStartScheduler();
+
+	// this function should never get here, nevertheless, please make sure that this last call doesn't get optimized away
+	exit(0);
```

Comment:
Is there a way to tell C to not optimize something away? I know in Verilog you would use something called `(no_prune)`

Reply:
There is the `volatile` keyword, although I don't know about its effect on function calls.

This call isn't actually very important so I'm not sure why it's here, the program would behave the same if it ever reached the end of the function, whether `exit(0)` was there or not (the behaviour is that it would simply reset).